pkg/sysregistriesv2/alias.go

```diff
@@ -0,0 +1,112 @@
+package sysregistriesv2
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/BurntSushi/toml"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/homedir"
+	"github.com/containers/storage/pkg/lockfile"
+)
+
+// userAliasesFile is the user-specific registries.conf file to store aliases.
+var userAliasesFile = filepath.FromSlash(".config/containers/aliases.conf")
+
+// userAliasesLockFile is the lockfile for the user-specific registries.conf
+// file to store aliases.
+var userAliasesLockFile = filepath.FromSlash(".local/share/containers/aliases.conf.lock")
+
+// ResolveAlias performs an alias resolution of the specified name.  The
+// user-specific aliases.conf has precedence over aliases in `conf`.
+func (conf *V2RegistriesConf) ResolveAlias(ctx *types.SystemContext, name string) (string, bool, error) {
```

Comment:
Could this return a `reference.Named`? I suspect inadvertent mixing of `string` values that all mean different things are a contributing factor to several of the semantic surprises in interpreting user input.

Reply:
I am torn as I enjoy the thought of preventing certain mistakes (and giving early feedback to the user in case they provided garbage in the conf) but it comes at cost of parsing each value when loading the config. If the list grew big, this could be a noticeable.

I have no feeling though how big the amount of aliases most be to have a noticeable performance impact. What's your take? I may be off.

Reply:
(I’m not worried about string copies/splits/… until given evidence to the contrary.)

`docker/reference` parsing is basically “regular expression” execution. I suspect those aren’t O(input size), but still probably good enough to support hundreds or thousands of values — especially with caching.

On my system, a `ParseNamed` takes around 3 microseconds (vs. 0.1 us for a `strings.Split`)

Reply:
… actually, _at this point_ it does not matter. We might argue about the cost of validating all entries when reading the full config files into memory, but validating a _single_ entry in `ResolveAlias`/`AddAlias` is clearly cheap enough.

Reply:
Just one is cheap but if we validate here, I also expect us to validate the configs to notify early on config errors.

Reply:
I wonder if it would make sense to have `name` be `reference.Reference`? It may avoid certain pitfalls _and_ `pkg/shortnames` will need to know if a user input includes a tag or digest and behave accordingly.

FWIW, I changed all return values to reference.Named

Reply:
`reference.Reference` is a weaker type (it only provides .`String()`). `reference.Named` is _at least_ a repo name, and the `reference.Named` value commonly (as returned by `ParseNormalizedNamed`) includes a tag/digest, and that’s reachable by an extra type check.

Reply:
I don't think we can use a `reference.Named` for the input as it may not include a repository. It could very well just be `foo` or `foo:tag`.

Reply:
Argh...it doesn't make sense here but rather in `pkg/shortnames`.

Reply:
Oops, you were asking about the _input_ to the lookup.

In my mind (which has _nothing_ to do with the actual software), only “normalized” values are `reference.Named`; short names are not (in Podman they are `imageParts`).

Basically, what I was asking for is to make the “possibly short name”, “certainly short name”, and “fully-qualified name” situations to be expressed by different Go types — certainly at least the fully-qualified from the other two. `string` vs. `reference.Named` is one way to do that — not at all an obvious way, admittedly.